### PR TITLE
CMake code refactor

### PR DESCRIPTION
**Pull request by mensinda**
This PR basically moves most of the CMake detection/parsing code from the dependency module into the CMake module. This change reduces the complexity of the `CMakeDependency` class and makes reusing the parser in the future much easier.

### CODE REVIEW

File: mesonbuild/cmake/traceparser.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        #
+        #   1. Assume that the property will be capitalized, this is convention
+        #      but cmake doesn't require it.
+        #   2. Maintain a copy of the list here: https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#target-properties
+        #
+        # Neither of these is awesome for obvious reasons. I'm going to try
+        # option 1 first and fall back to 2, as 1 requires less code and less
+        # synchroniztion for cmake changes.
+
+        arglist = []  # type: typing.List[typing.Tuple[str, typing.List[str]]]
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/traceparser.py]: __[Flake8]__ [F821] undefined name 'typing' ([link](https://sider.review/gh/repos/19784232/pulls/5457#issue-4181575))

<sub>posted by Sider</sub>

File: mesonbuild/cmake/executor.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+                if not silent:
+                    mlog.log('Found CMake:', mlog.red('NO'))
+                # Set to False instead of None to signify that we've already
+                # searched for it and not found it
+                CMakeExecutor.class_cmakebin[for_machine] = False
+                CMakeExecutor.class_cmakevers[for_machine] = None
+
+        return CMakeExecutor.class_cmakebin[for_machine], CMakeExecutor.class_cmakevers[for_machine], for_machine
+
+    def check_cmake(self, cmakebin: 'ExternalProgram') -> Optional[str]:
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/executor.py]: __[Flake8]__ [F821] undefined name 'ExternalProgram' ([link](https://sider.review/gh/repos/19784232/pulls/5457#issue-4181573))

<sub>posted by Sider</sub>

File: mesonbuild/cmake/interpreter.py [hunk 3]
```diff
@@ -263,7 +260,7 @@ def log(self) -> None:
             mlog.log('    -', key, '=', mlog.bold(str(val)))
 
 class CMakeInterpreter:
-    def __init__(self, build: Build, subdir: str, src_dir: str, install_prefix: str, env: Environment, backend: Backend):
+    def __init__(self, build: 'Build', subdir: str, src_dir: str, install_prefix: str, env: Environment, backend: 'Backend'):
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/interpreter.py]: __[Flake8]__ [F821] undefined name 'Build' ([link](https://sider.review/gh/repos/19784232/pulls/5457#issue-4181574))

<sub>posted by Sider</sub>

File: mesonbuild/cmake/executor.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+# This class contains the basic functionality needed to run any interpreter
+# or an interpreter-based tool.
+
+from .. import mlog, mesonlib
+from ..mesonlib import PerMachine, Popen_safe, version_compare
+from ..environment import Environment, MachineChoice
+
+from typing import List, Tuple, Optional, TYPE_CHECKING
+from subprocess import Popen, PIPE
```
Comment (ignatenkobrain) [inline mesonbuild/cmake/executor.py]: __[Flake8]__ [F401] 'subprocess.Popen' imported but unused ([link](https://sider.review/gh/repos/19784232/pulls/5457#issue-4181889))

<sub>posted by Sider</sub>

**Comment by jpakkane** [comment]
Can you rebase this to current master so that the CI gets rerun? Just to be sure, since the CMake code has changed in master.

**Pull request merged**